router.go

```diff
@@ -246,20 +263,64 @@ func (r *Router) putToS3(ctx context.Context, dest destination, body io.ReadSeek
 	r.sem.Acquire(ctx, 1)
 	defer r.sem.Release(1)
 
+	s3c, err := r.s3Client(ctx, dest.Bucket)
+	if err != nil {
+		return err
+	}
+
 	in := &s3.PutObjectInput{
 		Bucket:   &dest.Bucket,
 		Key:      &dest.Key,
 		Body:     body,
 		Metadata: meta,
 	}
 	log.Println("[info] starting put to", dest.String())
-	_, err := r.s3.PutObject(ctx, in)
-	if err == nil {
+	if _, err := s3c.PutObject(ctx, in); err == nil {
 		log.Println("[info] completed put to", dest.String())
 	}
 	return err
 }
 
+func (r *Router) defaultS3Client() *s3.Client {
+	return r.s3[r.awsConf.Region]
+}
+
+// s3Client returns s3 client for same region as the bucket
+func (r *Router) s3Client(ctx context.Context, bucket string) (*s3.Client, error) {
+	bucketRegion, err := r.getS3BucketRegion(ctx, bucket)
+	if err != nil {
+		return nil, err
+	}
+
+	if s3, ok := r.s3[bucketRegion]; ok {
+		return s3, nil
+	}
+
+	awsConfig := r.awsConf.Copy()
+	awsConfig.Region = bucketRegion
+	s3 := s3.NewFromConfig(awsConfig)
+	r.s3[bucketRegion] = s3
+
+	return s3, nil
+}
+
+func (r *Router) getS3BucketRegion(ctx context.Context, bucket string) (string, error) {
+	if region, ok := r.s3bucketRegion[bucket]; ok {
```

Comment:
`putToS3` might be called from multiple goroutines in parallel, so we have to guard concurrent map access. Please consider using `sync.Mutex` or another lock mechanism.